src/lib/Imp.hs

```diff
@@ -54,53 +56,58 @@ import Util
 --
 -- TODO: Use an ImpAtom type alias to better document the code
 
--- TODO: use `Env ()` instead of `Scope`. The problem is that we're mixing up
--- Imp and Core types in there.
-type EmbedEnv prog = ((Env Dest, [IVar]), (Scope, prog))
-type ImpMP prog = Cat (EmbedEnv prog)
-type ImpM = ImpMP ImpProgram
-
-toImpFunction :: ([ScalarTableBinder], Block) -> (ImpFunction, Atom)
-toImpFunction (vsIn, block) = runImpMPBinders @ImpProgram vsIn $ do
-  ((vsOut, atomOut), prog) <- scopedBlock $ materializeResult
-  return $ (ImpFunction (fmap toBinder vsOut) vsIn prog, atomOut)
-  where
-    toBinder = Bind . (\(Var x) -> x) . toArrayAtom . IVar
-    materializeResult = do
-      outDest <- allocKind Unmanaged $ getType block
-      void $ translateBlock mempty (Just outDest, block)
-      atomOut <- destToAtomScalarAction (return . toArrayAtom . IVar) outDest
-      return (destArrays outDest, atomOut)
-
-toMDImpFunction :: ([ScalarTableBinder], Block) -> (MDImpFunction ImpKernel, Atom)
-toMDImpFunction (vsIn, block) = runImpMPBinders @(MDImpProgram ImpKernel) vsIn $ do
-  ((vsOut, atomOut), prog) <- scopedBlock $ materializeResult
-  return $ (MDImpFunction (fmap toBinder vsOut) vsIn prog, atomOut)
+data ParallelismLevel = TopLevel | ThreadLevel  deriving (Show, Eq)
+data ImpCtx = ImpCtx { impBackend :: Backend
+                     , curDevice  :: Device
+                     , curLevel   :: ParallelismLevel }
+data ImpCatEnv = ImpCatEnv
+  { envPtrsToFree :: [IExpr]
+  , envScope      :: Scope
+  , envStatements :: [ImpStatement]
+  , envFunctions  :: Env ImpFunction }
+
+type ImpM = ReaderT ImpCtx (Cat ImpCatEnv)
+type AtomRecon = Abs (Nest Binder) Atom
+
+toImpModule :: Backend -> Block -> (ImpModule, [LitVal], AtomRecon)
```

Comment:
How about we parametrize this by the name of the `main` function? This way we wouldn't have to do the awkward pattern matching in LLVM to replace it with `entryFun`

Reply:
you mean the `Right (JIT.JITSymbol f _)` pattern? Not sure how that would work

Reply:
Oh, I meant the `impMainFunction` constant could be an argument to this function.